pkg/s3select/output.go

```diff
@@ -0,0 +1,623 @@
+var contHeaders []byte
+
+// Stat Headers
+// -11 -event type - 7 - 5 "Stat" -20
+// -13 -content-type -7 -8 "text/xml" -25
+// -13 -message-type -7 -5 "event"     -22
+// This is predefined from AMZ protocol found here:
+// https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html
+var statHeaders []byte
+
+// Progress Headers
+// -11 -event type - 7 - 8 "Progress" -23
+// -13 -content-type -7 -8 "text/xml" -25
+// -13 -message-type -7 -5 "event"     -22
+// This is predefined from AMZ protocol found here:
+// https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html
+var progressHeaders []byte
+
+// The length of the nonvariable portion of the ErrHeaders
+// The below are the specifications of the header for a "error" event
+// -11 -error-code - 7 - DEFINED "DEFINED"
+// -14 -error-message -7 -DEFINED "DEFINED"
+// -13 -message-type -7 -5 "error"
+// This is predefined from AMZ protocol found here:
+// https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html
+var errHdrLen int
+
+func init() {
+	recordHeaders = writeRecordHeader()
+	endHeaders = writeEndHeader()
+	contHeaders = writeContHeader()
+	statHeaders = writeStatHeader()
+	progressHeaders = writeProgressHeader()
+	errHdrLen = 55
+
+}
+
+// writePayloadSize writes the 4byte payload size portion of the protocol.
+func writePayloadSize(payloadSize int, headerLength int) []byte {
+	totalByteLen := make([]byte, 4)
+	totalMsgLen := uint32(payloadSize + headerLength + 16)
+	binary.BigEndian.PutUint32(totalByteLen, totalMsgLen)
+	return totalByteLen
+}
+
+// writeHeaderSize writes the 4byte header size portion of the protocol.
+func writeHeaderSize(headerLength int) []byte {
+	totalHeaderLen := make([]byte, 4)
+	totalLen := uint32(headerLength)
+	binary.BigEndian.PutUint32(totalHeaderLen, totalLen)
+	return totalHeaderLen
+}
+
+//  writeCRC writes the CRC for both the prelude and and the end of the protocol.
+func writeCRC(myBuffer []byte) []byte {
+	// Calculate the  CRC here:
+	myCRC := make([]byte, 4)
+	cksum := crc32.ChecksumIEEE(myBuffer)
+	binary.BigEndian.PutUint32(myCRC, cksum)
+	return myCRC
+}
+
+// writePayload writes the Payload for those protocols which the Payload is
+// necessary.
+func writePayload(myPayload string) []byte {
+	convertedPayload := []byte(myPayload)
+	payloadStore := make([]byte, len(convertedPayload))
+	copy(payloadStore[0:], myPayload)
+	return payloadStore
+}
+func writeRecordHeader() []byte {
+	var currentMessage = &bytes.Buffer{}
+
+	headerNameBytes := make([]byte, 1)
+	headerNameBytes[0] = 11
+	currentMessage.Write(headerNameBytes)
+
+	headerName := make([]byte, 11)
+	copy(headerName[0:], ":event-type")
+	currentMessage.Write(headerName)
+
+	headerType := make([]byte, 1)
+	headerType[0] = 7
+	currentMessage.Write(headerType)
+
+	headerNameSize := make([]byte, 2)
+	binary.BigEndian.PutUint16(headerNameSize, 7)
+	currentMessage.Write(headerNameSize)
```

Comment:
All the write*Header() functions seem to be quite tedious to read - you can make this better by writing some helper functions and combining strings with `append()`. E.g:

```go
// Encode s in a []byte, lenBytes is the number of bytes used to encode the length of the string.
func encodeString(s string, lenBytes int) []byte {
        n := len(s)
        lenSlice := make([]byte, lenBytes)
        binary.BigEndian.PutUint16(lenSlice[0:], n)
        return append(lenSlize, []byte(s))
}

func encodeNumber(n int64, lenBytes int) {
        lenSlice := make([]byte, lenBytes)
        binary.BigEndian.PutUint16(lenSlice[0:], n)
        return lenSlice
}
```

With the above, `writeRecordHeader()` would just look something like:

```go
        return append(nil,
                  // header name
                  encodeSlice(":event-type", 1),
                  // header type
                  encodeNumber(7, 1)
                 // header name size
                  encodeNumber(7, 2),
                // etc etc
```